### ISSUE TITLE
Dependency Management, Including Modules, and Test Doubles

### ISSUE THREAD

**Issue by mikewacker** [dagger/dagger#5715]
### What is the issue?

This is a topic that the existing [testing guidance](https://dagger.dev/dev-guide/testing.html) doesn't cover.

**tl;dr:** I want modules to include their production dependencies by default, but I still want flexibility to use test doubles.

---

Let's use the following example to explain this in more depth:

- We have three public interfaces (each in a different package): `Server`, and `Database`, and `Config`.
- We have three corresponding package-private implementations: `ServerImpl`, `DatabaseImpl`, and `ConfigImpl`.
- We have three corresponding public modules: `ServerModule`, `DatabaseModule`, and `ConfigModule`.
    - Each module has one published binding. E.g., `ServerModule` binds `ServerImpl` to `Server`.
    - Each module may have additional internal bindings. 
- `ServerImpl`'s `@Inject`'ed constructor includes a `Database` arg, a `Config` arg, and possibly other internal args.
- `DatabaseImpl`'s `@Inject`'ed constructor includes a `Config` arg and possibly other internal args.

---

Let's assume that we only have to worry about creating a production `Server`.

In that case, ideally `ServerModule` would include its dependencies:

```java
@Module(includes = {DatabaseModule.class, ConfigModule.class})
public interface ServerModule {

    @Binds
    Server bindServer(ServerImpl impl);

    // other internal bindings
}
```

Intuitively, if we're only worrying about production, it does make sense to have `ServerModule` manage its own dependencies&mdash;rather than make a (sub/)`Component` manage the transitive dependencies for `ServerModule`.

Here's a simplified `Component`:

```java
@Component(modules = ServerModule.class)
@Singleton
public interface ServerComponent {

    Server server();
}
```

---

Next, let's assume that I want to write tests for `Server` (backed by a `ServerImpl`) that uses a test double for `Database`: a stub database, an in-memory database, etc.

How do we do that? This is the best answer I've come up with so far:

```java
@Module(includes = {ServerModule.NoDeps.class, DatabaseModule.class, ConfigModule.class})
public interface ServerModule {

    @Module
    interface NoDeps {

        @Binds
        Server bindServer(ServerImpl impl);

         // other internal bindings
    }
}
```

Here's how it would work:

- For production, nothing changes; you would still include `ServerModule`.
- For testing, if you want different dependencies, you can include `ServerModule.NoDeps` instead.

E.g.:

```java
@Module(includes = {ServerModule.NoDeps.class, InMemoryDatabaseModule.class, ConfigModule.class})
public interface TestServerModule {}
```

Are there any better options than this?

**Comment by wingyplus** [comment]
This is repository for dagger, the ci/cd as code.  Not the dagger, dependency injection tool. :)

**Comment by mikewacker** [comment]
Whoops...
